### ISSUE TITLE
consider adding JDBC type to @Column or a zoo of built-in AttributeConverters

### ISSUE THREAD

**Issue by gavinking** [jakartaee/persistence#619]
Currently there are two ways to customize the column type mapped by a field:

1. using an `AttributeConverter`, or
2. using `@Column(columnDefinition)`.

An `AttributeConverter` works very well for basic types, but there are no built-in `AttributeConverter`s, and it's a pain for the user to have to write one. And it doesn't work at all for associations.

On the other hand `columnDefinition` doesn't work very well, for several reasons, including that it's not at all portable between databases, and also doesn't do any sort of type conversion at the Java level.

First approach
---

Since very ancient times, Hibernate allowed the column type to be customized in a portable way by specifying a JDBC type code from the list given in `SqlTypes`. Today this is usually done via the `@JdbcTypeCode` annotation.

The only thing about this approach that has always left me a little uncomfortable is the lack of typesafety, due to the `SqlTypes` being integers. A way to alleviate this discomfort would be to use the `JDBCType` enumeration instead. So we could add:

```java
public @interface Column {
    ...
    JDBCType type() default JDBCType.OTHER;
}
```

where we would be abusing `OTHER` to mean "default".

So I could write:

```java
@Column(type=JDBCType.DECIMAL)
int total;
```

The spec would need to define which combinations of (field type, JDBC type) an implementation is required to support, but just like with the JPQL `cast()` function, that list need not be extremely long. That is, it could, at least initially, just be basic numeric conversions and conversions to `varchar`.

Second approach
---

Alternatively, if this abuse of `OTHER` is unacceptable, we could add a new annotation:

```java
public @interface ColumnType {
    JDBCType value();
}
```

which might anyway be a cleaner approach.

I would write:

```java
@ColumnType(JDBCType.DECIMAL)
int total;
```

Third approach
---

Alternatively, the specification could provide a set of built-in attribute converter classes, for use with `@Convert`. The problem is that to provide converters for _n_ Java types to _m_ JDBC types requires _nm_ classes, which gets a bit messy.

It would look like:

```java
@Convert(converter=Converters.IntegerToDecimal.class)
int total;
```

which is a bit more verbose than the previous options.

The advantage of this approach is that it requires no new concept in the spec.

**Comment by gavinking** [comment]
Of course, there's also a

Fourth approach
----

which is just to do what Hibernate does today and use `int`-valued codes from `Types`.

```java
public @interface Column {
    ...
    int type() default Integer.MIN_VALUE;
}
```

letting me write:

```java
@Column(type=Types.DECIMAL)
int total;
```

The payoff for the loss of typesafety is that the set of allowed type codes is provider-extensible, which is definitely useful.

**Comment by beikov** [comment]
I think both, the first and second approach are ok, though the first one seems easier as the second might also require the addition of `@JoinColumnType` etc.

The fourth option is provider extensible, but I think we should for provider specifics, it's best to stick to custom annotations.

Also, I don't think that `JDBCType` is a fitting name, since what is being described here is an abstract SQL type.

**Comment by gavinking** [comment]
> Also, I don't think that `JDBCType` is a fitting name, since what is being described here is an abstract SQL type.

Well, OK, but it's `java.sql.JDBCType` and we can't change that (and I would not want to introduce a new `enum` isomorphic to it either).

**Comment by beikov** [comment]
Ah, you were referring to the JDK enum. I guess it's ok to reuse that, though that limits JPA to the types JDBC vendors agree on. I think I prefer that we leave the door open to allow extensions that go beyond the JDBC spec.

**Comment by beikov** [comment]
I would like to mention that array support is somewhat "special" in the sense that Hibernate ORM (and by extension probably other JPA providers) need to know the element type to produce properly typed results for array functions.
In the case of JSON, we might want to differentiate between `json array` i.e. `json[]` representing a SQL array of json values and `json_array` i.e. `json` representing a json value that contains an array.

So maybe it's better to move arrays out of this equation and instead consider a `@Array` annotation like we have in Hibernate ORM.

**Comment by gavinking** [comment]
Yes but that's one of the advantages of using a string here. The user can write `int array` or whatever.

**Comment by gavinking** [comment]
But note that this is _not_ a proposal to add support for arrays or JSON. It's only about how to replace `columnDefinition` with something portable if, for example, you want to store a `float` in a `numeric` column or whatever.

**Comment by gavinking** [comment]
So I had a little time to reflect a bit more on this one,

What I notice about this is a don't hate *any* of the four ways I proposed, to the extent that when I consider any one of them I think, "oh yeah, that's fine". On the other hand I just don't love any of them either.

Ultimately, to make any one of these things worth doing and making a requirement of the spec means threading the needle between existing options. The new thing has to be clearly better than:

- `@Column(columnDefinition= ...)` (not very hard)
- `@Convert(converter=MyAttributeConverter.class)`
- vendor-specific options (like, in Hibernate, `@JdbcType(VARCHAR)` or `@Type(MyUserType.class)`)

and it's just really difficult to come up with something like that. Each of the options above has its own little sweet spot, but the sweet spot just isn't very big, when it's crowded out by the already-existing possibilities.

So I ended up with the conclusion that JPA's solution to this problem is the `@Convert` annotation, and we're not going to introduce a "new thing".

However, there is an incredibly simple enhancement we could make to `@Convert` that:

1. is completely type safe
2. is somewhat extensible
3. is portable for the simple cases
4. has an extremely small blast radius 

and that is, ta-da, just add a `to` member to `@Convert`:

```java
@Convert(to=BigDecimal.class)
int total;
```

We would require that implementations support basic numeric conversion, as well as certain conversions to and from String. And, of course, implementations would be free to support other built-in conversions going beyond those.

Being just a shortcut sort of `AttributeConverter`, this is extremely easy to specify. The only downside of this is the thing I mentioned above that an `AttributeConverter` isn't allowed on an `@Id`. But in principle we could relax that for these built-in conversions.

**Comment by gavinking** [comment]
I have implemented a more complete proposal in #819.
